ui/src/App.js

```diff
@@ -1,108 +1,382 @@
+    connect: false,
+    tray_reset_account: false,
+    show_secret_phrase: false
+  });
+  const [exportedSecretPhrase, setExportedSecretPhrase] = useState(null);
+  const [exportingPhrase, setExportingPhrase] = useState(false);
+  const exportingPhraseRef = useRef(exportingPhrase);
+  const pathnameRef = useRef(location.pathname);
+
+  useEffect(() => {
+    pathnameRef.current = location.pathname;
+  }, [location.pathname]);
 
   useEffect(() => {
     if (isConnected) return;
     const beginInitialConnectionPhase = async () => {
-      await once('connect', (event) => {
-        invoke('ui_connected');
+      await listen('connect', (event) => {
+        invoke('connect_ui');
         console.log("[INFO]: Connect Event: ", event);
         let payload = event.payload;
+
+        // We don't want to switch the page on reset during recovery process.
+        if (pathnameRef.current === ("/recover/loading")) {
+          setPayloadType('CreateAccount');
+          setRecoveryPhrase(payload.content);
+          return;
+        }
+
         switch (payload.type) {
           case 'CreateAccount':
             setRecoveryPhrase(payload.content);
-            setCurrentPage(CREATE_ACCOUNT_PAGE);
+            setPayloadType('CreateAccount');
+            navigate("/create-or-recover");
             break;
           case 'Login':
-            setCurrentPage(LOGIN_PAGE);
+            setPayloadType('Login');
+            invoke('enable_reset_menu_item');
+            navigate("/sign-in");
             break;
           default:
             break;
         }
       });
     };
-    beginInitialConnectionPhase();
-  }, [isConnected]);
+
+    const listenForResetTrayRequests = async () => {
+      console.log("[INFO]: Setup tray reset listener.");
+      listen('tray_reset_account', async (event) => {
+        console.log("[INFO]: Wake: ", event);
+
+        // checking to make sure to end password stalling listener on backend if we switch away
+        // from the export secret phrase.
+        if (exportingPhraseRef.current) {
+          console.log("[INFO]: Ending export phrase process.");
+          await stopPasswordPrompt();
+          endExportPhrase(false);
+        }
+        navigate("/reset");
+      })
+    }
+
+    if (!activeListeners.connect) {
+      beginInitialConnectionPhase();
+      setActiveListeners({
+        ...activeListeners,
+        connect: true,
+      });
+    }
+    if (!activeListeners.tray_reset_account) {
+      listenForResetTrayRequests();
+      setActiveListeners({
+        ...activeListeners,
+        tray_reset_account: true,
+      });
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [isConnected, activeListeners, pathnameRef, navigate]);
+
+  // keeps show secret phrase listener in sync with exportingPhrase state
+  // whether or not we are currently exporting the phrase.
+  useEffect(() => {
+    exportingPhraseRef.current = exportingPhrase;
+  }, [exportingPhrase])
```

Comment:
I agree, this useEffect was totally not needed, I only needed the useRef because the listener is unable to get the latest value of a useState variable. I have removed the useEffect and useState and simplified it to used only the useRef.